### PR TITLE
Fix the composer to propagate nuxt/vue errors back & Add customizable hooks & Add better debug info

### PR DESCRIPTION
**Pull request by FieryCod**
This PR fixes the bug which we encountered with @kamil-stasiak when we're trying to develop the fetch more button. Simply but the error was that when some of `vue.js/nuxt.js` error occurred the composer did not propagate the error back. 

When it comes to changes in the composer I tried to make it more readable. I introduced the errors folder which is used to create new errors. Then I can do some `if` magic and call different functions on their occurrence.

I was thinking about the use of `instanceof` instead of `err.name` but it turned out that it needs some custom polyfill in `Babel` so because I wanted the module to be 0 dependant I just used the `err.name`.

I do not know if the `callers` section is needed. I'm trying to figure it out. I think that I gonna wait couple of `PR's` and if the usage of caller would be limited to only one I gonna remove it.

`JsDoc's` will be added in future `PR's`. 

Ekhm.. I have almost forgotten about one thing. The logs. Debug logs are now beautiful. I'm very proud of it. You definitely must check them out.

### CODE REVIEW

**Comment by k911** [comment]
Once again, when you feel that you should add "AND" in commit message, it means that commit should be sparated to serval commits

File: helpers/store/composer.js [hunk 1]
```diff
@@ -1,43 +1,89 @@
...
       } catch (err) {
-        // When something is canceled, or when action does not succeeds and all fails have been called
+        // When error occurs or some custom hooks
+
+        if (err.name === 'ActionFail' && !_.isEmpty(fail)) {
+          // When action failed
+
+          knitLogger.debug(() => 'Calling on fail functions')
+          await _promisesBoundCaller(fail, this, ctx, params, null)
+        } else if (err.name === 'ActionCancel' && !_.isEmpty(cancel)) {
```
Comment (k911) [inline helpers/store/composer.js]: Please use consts for 'ActionCancel' etc.

File: helpers/store/errors/ActionFail.js [hunk 2]
```diff
@@ -0,0 +1,6 @@
+export default class ActionFail extends Error {
+  constructor (actionName) {
+    super(`Action failed: ${actionName}`)
+    this.name = 'ActionFail'
```
Comment (k911) [inline helpers/store/errors/ActionFail.js]: const

File: store/view/articles.js [hunk 3]
```diff
@@ -50,11 +50,12 @@ export const state = () => ({
 // Module actions
 export const actions = {
   getMainPage: composer.compose({
+    name: '#getMainpage',
```
Comment (k911) [inline store/view/articles.js]: const, and should be "#getMainPage" as stated before main page is two separate words

**Comment by FieryCod** [comment]
@k911 You're right it's really last time I'm doing that. I should remember that commit is per one meaningful change. Thanks for giving me nice hints about code standards.

File: helpers/store/composer.js [hunk 4]
```diff
@@ -1,43 +1,89 @@
+import commonHelper from '~/helpers/common'
+import knitLogger from '~/config/logger'
 import _ from 'lodash'
 
-const methods = {
-  cancel () {
-    throw new Error('cancel')
+function _validateComposableInputs (name, caller) {
```
Comment (k911) [inline helpers/store/composer.js]: Why do you use underscore now as name? I suppose you try to express it as internal, should not be exported? Is it really needed? Ping @zrebcu411

File: helpers/store/composer.js [hunk 5]
```diff
@@ -1,43 +1,89 @@
...
-  // Binds all parameters to all composers, caller
-  const flowWithBoundCaller = (that, ctx, params) => {
-    const beforeBound = _.flow(before).bind({ that, ctx, params, methods })
-    return _.flow([ beforeBound, caller.bind({ that, ctx, params, methods: { ...methods, promisesBoundCaller }, fail }) ])
-  }
+  return _.flow([ beforeBound, caller.bind(bound) ])
+}
+
+function compose ({ name, before, caller, success, always, fail, cancel }) {
+  const programaticErrors = ['ActionFail', 'ActionCancel']
```
Comment (k911) [inline helpers/store/composer.js]: const

File: helpers/store/errors/ActionCancel.js [hunk 6]
```diff
@@ -0,0 +1,6 @@
+export default class ActionCancel extends Error {
+  constructor (actionName) {
+    super(`Action canceled: ${actionName}`)
+    this.name = 'ActionCancel'
```
Comment (k911) [inline helpers/store/errors/ActionCancel.js]: const

File: store/view/articles.js [hunk 7]
```diff
@@ -70,13 +71,14 @@ export const actions = {
     }, { root: true })
   }),
   getCategoryList: composer.compose({
+    name: '#getCategory',
```
Comment (k911) [inline store/view/articles.js]: const

File: store/view/articles.js [hunk 8]
```diff
@@ -92,12 +94,13 @@ export const actions = {
     }, { root: true })
   }),
   getMainList: composer.compose({
+    name: '#getMainlist',
```
Comment (k911) [inline store/view/articles.js]: const and should be "#getMainList"

File: store/view/articles.js [hunk 9]
```diff
@@ -113,11 +116,12 @@ export const actions = {
     }, { root: true })
   }),
   getArticle: composer.compose({
+    name: '#getArticle',
```
Comment (k911) [inline store/view/articles.js]: const

File: helpers/store/composer.js (hunk 4 above)
Comment (FieryCod) [reply helpers/store/composer.js]: Yes I'm trying to articulate that it's internal and should not be exported. I think yes it's needed and we should validate the inputs in order to create logs, and act properly.

File: helpers/store/compounds.js [hunk 10]
```diff
@@ -6,14 +7,17 @@ function callActionWhen (predicate) {
...
-    knitLogger.debug(() => `Predicate: ${JSON.stringify(predicate)} evaluated to false, no action called!`)
-    this.methods.cancel()
+    knitLogger.debug(() => `Predicate: ${predicate} evaluated to false, canceling action!`)
+
+    // Cancel action
+    throw new ActionCancel(this.that.actionName)
   }
 }
 
 function doSth (fn) {
```
Comment (k911) [inline helpers/store/compounds.js]: Wow, how do I miss naming this function as it is? :( Please change it to proper name, and add description what it should be used for. BTW Is it really needed?

File: helpers/store/composer.js [hunk 11]
```diff
@@ -1,43 +1,89 @@
...
+}
+// Binds all parameters to the action
+const _actionWithBoundCaller = (that, ctx, params, actionDefn) =>
+  () => actionDefn.bind(that)(ctx, params)
 
-  // Binds all parameters to the action
-  const actionWithBoundCaller = (that, ctx, params, actionDefn) =>
-    () => actionDefn.bind(that)(ctx, params)
+// Binds all parameters to all composers, caller
+const _flowWithBoundCaller = (before, caller, that, ctx, params) => {
```
Comment (k911) [inline helpers/store/composer.js]: `that`, `ctx`, `params` are duplicated in `_actionWithBoundCaller` and `_flowWithBoundCaller` maybe we should think to be consistent and have them in the same order in declaration?

i propose:

'const _actionWithBoundCaller = (that, ctx, params, actionDefn)`
and
`const _flowWithBoundCaller = (that, ctx, params, before, caller)`

File: helpers/store/callers.js [hunk 12]
```diff
@@ -1,22 +1,23 @@
...
 import knitLogger from '~/config/logger'
 
-function squashErrorsCall (message = '') {
-  return function (action) {
-    return Promise
-      .resolve(action())
-      .catch(async (err) => {
-        knitLogger.debug(() => err)
-        knitLogger.debug(() => message)
+// NOTE: I think that there will be only one caller but let's wait couple of PR's
```
Comment (k911) [inline helpers/store/callers.js]: no need for that note, please delete

**Comment by FieryCod** [comment]
Wow guys you're so good for me I really appreciate your help, your feedback and your nice ideas which make code simpler, easier to reason and so on. Much thanks.

**Pull request merged**